lib/resty/openssl/x509/csr.lua

```diff
@@ -162,19 +166,73 @@ function _M:to_PEM()
   return tostring(self, "PEM")
 end
 
+--- Get all csr extensions
+-- @tparam table self Instance of csr
+-- @treturn List of parsed extension objects
+function _M.get_extensions(self)
+    local extensions = C.X509_REQ_get_extensions(self.ctx)
+    local n = stack_macro.OPENSSL_sk_num(extensions)
+    local ret = {}
+    for i = 0, n - 1 do
+        local ext = stack_macro.OPENSSL_sk_value(extensions, i)
+        local dup, err = extension_lib.dup(ffi_cast(ext_typ_ptr, ext))
+        if not err then
+            local obj = dup:get_object()
+            obj.blob = dup:tostring()
+            push(ret, obj)
+        end
+    end
+    return ret
+end
+
+--- Get a csr extension
+-- @tparam table self Instance of csr
+-- @tparam string|number Nid number or name of the extension
+-- @treturn Parsed extension object or nil if not found
+function _M.get_extension(self, nid)
+    local i, err = txtnid2nid(nid)
+    if err then
+        return nil, err
+    end
+    local items = self:get_extensions()
+
+    for j = 1, #items do
+        local ext = items[j]
+        if ext.nid == i then
+            return ext
+        end
+    end
+    return nil, ("extension for %d not found"):format(nid)
+
+    -- @todo fix malloc issue and use code below
+    --[[    local ctx = C.X509V3_EXT_i2d(i, 0, self.ctx)
+        if ctx == nil then
+            return nil, format_error("csr.get_extension: X509V3_EXT_i2d")
+        end
+        ffi_gc(ctx, C.X509_EXTENSION_free)
+        local dup = extension_lib.dup(ctx)
+        local obj = dup:get_object()
+        obj.blob = dup:tostring()
+        return obj]]
+end
+
+
 -- START AUTO GENERATED CODE
 
 -- AUTO GENERATED
 function _M:sign(pkey, digest)
   if not pkey_lib.istype(pkey) then
     return false, "x509.csr:sign: expect a pkey instance at #1"
   end
-  if digest and not digest_lib.istype(digest) then
+  if not digest or not digest_lib.istype(digest) then
     return false, "x509.csr:sign: expect a digest instance at #2"
   end
+  if not digest.dtyp then
+    return false, "x509.csr:sign: expect a digest instance should have dtyp member"
+  end
 
   -- returns size of signature if success
-  if C.X509_REQ_sign(self.ctx, pkey.ctx, digest and digest.ctx) == 0 then
+  if C.X509_REQ_sign(self.ctx, pkey.ctx, digest and digest.dtyp) == 0 then
```

Comment:
ah this is a good fix 👍 since this function is generated, could you update it in the template instead? it's at https://github.com/fffonion/lua-resty-openssl/blob/master/scripts/templates/x509_functions.j2#L3

Reply:
Done